synapse/api/ratelimiting.py

```diff
@@ -13,79 +13,121 @@
+                * The time in seconds of when it can next be performed.
+                  -1 if a rate_hz has not been defined for this Ratelimiter
         """
-        self.prune_message_counts(time_now_s)
-        message_count, time_start, _ignored = self.message_counts.get(
-            key, (0.0, time_now_s, None)
-        )
+        # Remove any expired entries
+        self._prune_message_counts(time_now_s)
+
+        # Check if there is an existing count entry for this key
+        action_count, time_start, = self.actions.get(key, (0.0, time_now_s))
+
+        # Check whether performing another action is allowed
         time_delta = time_now_s - time_start
-        sent_count = message_count - time_delta * rate_hz
-        if sent_count < 0:
+        performed_count = action_count - time_delta * self.rate_hz
+        if performed_count < 0:
+            # Allow, reset back to count 1
             allowed = True
             time_start = time_now_s
-            message_count = 1.0
-        elif sent_count > burst_count - 1.0:
+            action_count = 1.0
+        elif performed_count > self.burst_count - 1.0:
+            # Deny, we have exceeded our burst count
             allowed = False
         else:
+            # We haven't reached our limit yet
             allowed = True
-            message_count += 1
+            action_count += 1.0
 
         if update:
-            self.message_counts[key] = (message_count, time_start, rate_hz)
+            self.actions[key] = (action_count, time_start)
 
-        if rate_hz > 0:
-            time_allowed = time_start + (message_count - burst_count + 1) / rate_hz
+        # Figure out the time when an action can be performed again
+        if self.rate_hz > 0:
+            time_allowed = (
+                time_start + (action_count - self.burst_count + 1) / self.rate_hz
+            )
+
+            # Don't give back a time in the past
             if time_allowed < time_now_s:
                 time_allowed = time_now_s
         else:
+            # This does not apply
             time_allowed = -1
 
         return allowed, time_allowed
 
-    def prune_message_counts(self, time_now_s):
-        for key in list(self.message_counts.keys()):
-            message_count, time_start, rate_hz = self.message_counts[key]
+    def _prune_message_counts(self, time_now_s: int):
+        """Remove message count entries that are older than
+
+        Args:
+            time_now_s: The current time
+        """
+        # We create a copy of the key list here as the dictionary is modified during
+        # the loop
+        for key in list(self.actions.keys()):
+            action_count, time_start = self.actions[key]
+
             time_delta = time_now_s - time_start
-            if message_count - time_delta * rate_hz > 0:
+            if action_count - time_delta * self.rate_hz > 0:
+                # XXX: Should this be a continue?
                 break
```

Comment:
I'm fairly certain that this is meant to be a continue. Did spelunking in the git blame show anything about this line?